src/data_readers/data_reader_image.cpp

```diff
@@ -136,6 +136,13 @@ void image_data_reader::load() {
   select_subset_of_data();
 }
 
+void image_data_reader::setup() {
+  generic_data_reader::setup();
+
+  using InputBuf_T = lbann::cv_image_type<uint8_t>;
+  m_thread_cv_buffer.resize(omp_get_max_threads(), cv::Mat(1, get_linearized_data_size(), InputBuf_T::T(1)));
```

Comment:
I'm not sure this is correct. The elements in `m_thread_cv_buffer` will be initialized by copying the input value. I believe the copy constructor for `cv::Mat` does not duplicate the underlying buffer, but instead creates a reference to it. (See [here](https://docs.opencv.org/3.4/d3/d63/classcv_1_1Mat.html#a294eaf8a95d2f9c7be19ff594d06278e).) So we probably want to explicitly create separate matrices for each thread.

Reply:
m_thread_cv_buffer seems to reserve the buffer space for each thread. 
However, my concern is the hard-coded type uint8_t. The entire pipeline has been designed to handle various types. Now, this breaks that.

Reply:
Since all the images we deal with using data_reader_image derivatives are of uint8_t, we can address this issue later. However, notice that there are other types of images. The image from jag simulation is floating point type although we do not use any data_reader_image derivative for that. I have dealt with X-ray images of 16bit/32bit integers in other project.

Reply:
`m_thread_cv_buffer` reserves a `cv::Mat` for each thread, yes. My concern is that all of those `cv::Mat`s share the same underlying buffer, unless this is reinitialized elsewhere and I've missed it.

Reply:
This loading creates a new buffer per image, and the subsequent preprocessing steps modifies the buffer in-place if possible. So, the buffers should rather be per minibatch item than per thread. If there is double buffering used, then per minibatch and per buffering stage.

Reply:
No, that's not the problem. Unless I've misunderstood something, this is what is going on: The problem is that every thread will share the same underlying buffer. The `std::vector` constructor being used (`vector(size_type count, const T& value)`) constructs the container by copying `value` once for each of the `count` entries. The copy constructor for `cv::Mat` does a shallow copy and just grabs a reference to the underlying buffer passed in. Hence, every element in that vector will point to the same underlying matrix.

Reply:
Oh, I see that!